lightning/src/ln/channelmonitor.rs

```diff
@@ -1801,22 +1808,49 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
-		// tracking state and panic!()ing if we get an update after force-closure/local-tx signing.
 		log_trace!(self, "Getting signed latest local commitment transaction!");
-		if let &mut Some(ref mut local_tx) = &mut self.current_local_signed_commitment_tx {
-			local_tx.tx.add_local_sig(&self.onchain_detection.keys.funding_key(), self.funding_redeemscript.as_ref().unwrap(), self.channel_value_satoshis.unwrap(), &self.secp_ctx);
+		if let Some(commitment_tx) = self.onchain_tx_handler.get_fully_signed_local_tx(self.channel_value_satoshis.unwrap()) {
+			let txid = commitment_tx.txid();
+			let mut res = vec![commitment_tx];
+			if let &Some(ref local_tx) = &self.current_local_signed_commitment_tx {
+				for htlc in local_tx.htlc_outputs.iter() {
+					if let Some(htlc_index) = htlc.0.transaction_output_index {
+						let preimage = if let Some(preimage) = self.payment_preimages.get(&htlc.0.payment_hash) { Some(*preimage) } else { None };
+						if let Some(htlc_tx) = self.onchain_tx_handler.get_fully_signed_htlc_tx(txid, htlc_index, preimage) {
+							res.push(htlc_tx);
+						}
+					}
+				}
+				// We throw away the generated waiting_first_conf data as we aren't (yet) confirmed and we don't actually know what the caller wants to do.
+				// The data will be re-generated and tracked in check_spend_local_transaction if we get a confirmation.
+			}
+			return res
 		}
-		if let &Some(ref local_tx) = &self.current_local_signed_commitment_tx {
-			let mut res = vec![local_tx.tx.with_valid_witness().clone()];
-			res.append(&mut self.broadcast_by_local_state(local_tx).0);
-			// We throw away the generated waiting_first_conf data as we aren't (yet) confirmed and we don't actually know what the caller wants to do.
-			// The data will be re-generated and tracked in check_spend_local_transaction if we get a confirmation.
-			res
-		} else {
-			Vec::new()
+		Vec::new()
+	}
+
+	/// Unsafe test-only version of get_latest_local_commitment_txn used by our test framework
+	/// to bypass LocalCommitmentTransaction state update lockdown after signature and generate
+	/// revoked commitment transaction.
+	#[cfg(test)]
+	pub fn unsafe_get_latest_local_commitment_txn(&mut self) -> Vec<Transaction> {
+		log_trace!(self, "Getting signed copy of latest local commitment transaction!");
+		if let Some(commitment_tx) = self.onchain_tx_handler.get_fully_signed_copy_local_tx(self.channel_value_satoshis.unwrap()) {
+			let txid = commitment_tx.txid();
+			let mut res = vec![commitment_tx];
+			if let &Some(ref local_tx) = &self.current_local_signed_commitment_tx {
+				for htlc in local_tx.htlc_outputs.iter() {
+					if let Some(htlc_index) = htlc.0.transaction_output_index {
+						let preimage = if let Some(preimage) = self.payment_preimages.get(&htlc.0.payment_hash) { Some(*preimage) } else { None };
+						if let Some(htlc_tx) = self.onchain_tx_handler.get_fully_signed_htlc_tx(txid, htlc_index, preimage) {
```

Comment:
This seems to violate the docs for sign_htlc_transacion, which says "If the transaction is an
    /// HTLC-Success transaction (ie htlc.offered is false), preimage must be set!"?

Reply:
Yes this is currently enforced in OnchainTxHandler::get_fully_signed_htlc_tx. IIRC I doubt a bit between checking in ChannelMonitor or OnchainTxHandler, I think that's the same and only temporary waiting for a further API refactor.

Reply:
Ah, I missed that you added a check in add_htlc_sig that didn't exist in sign_htlc_transaction without updating the comment. Its kinda awkward that the LocalCommitmentTransaction layer checks that the caller did something sane, but, for now, OK.